zerver/lib/mention.py

```diff
@@ -162,15 +171,52 @@ def get_stream_name_map(self, stream_names: Set[str]) -> Dict[str, int]:
                 .values(
                     "id",
                     "name",
+                    "recipient_id",
                 )
             )
 
             for row in rows:
                 self.stream_cache[row["name"]] = row["id"]
+                self.recipient_cache[row["name"]] = row["recipient_id"]
                 result[row["name"]] = row["id"]
 
         return result
 
+    def get_topic_info_map(
+        self, channel_topic: Set[ChannelTopicInfo]
+    ) -> Dict[ChannelTopicInfo, int]:
+        if not channel_topic:
+            return {}
+
+        result: Dict[ChannelTopicInfo, int] = {}
+        unseen_channel_topic: List[ChannelTopicInfo] = []
+
+        for channel_topic_object in channel_topic:
+            if channel_topic_object in self.topic_cache:
+                result[channel_topic_object] = self.topic_cache[channel_topic_object]
+            else:
+                unseen_channel_topic.append(channel_topic_object)
+
+        for channel_topic_object in unseen_channel_topic:
+            recipient_id = self.recipient_cache.get(channel_topic_object.channel_name)
+
+            assert recipient_id is not None
+
+            topic_latest_message = (
+                messages_for_topic(self.realm_id, recipient_id, channel_topic_object.topic_name)
+                .values_list("id", flat=True)
+                .order_by("-id")
+                .first()
+            )
+
+            if topic_latest_message is None:
+                continue
+
+            self.topic_cache[channel_topic_object] = topic_latest_message
+            result[channel_topic_object] = topic_latest_message
+
+        return result
```

Comment:
This logic is not correct for private channels with protected history. You can use this logic for channels with shared history, but otherwise, the query needs to be joined with `UserMessage`.

For channels with protected history, you'll need a function along the lines of `user_message_exists_for_topic` that returns the last message.

Reply:
oh ok, right. I haven't thought of protected history

Reply:
Technically, I don't think it would be big of a problem with having same logic for channels with protected history as well because -- even if I am able to `#-mention` a topic which I am totally unaware of after subscribing to such a channel, I still won't be able to access it because before fetching messages of that narrow, we first check if that message in `with` operand is accessible to user by `access_message`.

If you're concerned about displaying of message id in the link which is not accessible to users, then we should also worry about that display of message id when some other user mentions that link.

Reply:
One possible setback if we wish to use the `UserMessage` for this is that the markdown doesn't have any `user_profile` argument -- so possibly I need to make a prep commit to be able to do that?

Reply:
I updated and added some tests.